File: segment_tree/interesting_queries.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
public:
  vector<int> solveQueries(vector<int> nums, vector<vector<int>> Queries,
                           int k) {
    // Code here
    int n = nums.size();
    int nq = Queries.size();
    vector<int> ans(nq);
    int block_size = sqrt(n) + 1;
    vector<tuple<int, int, int>> query;

    for (int i = 0; i < nq; i++) {
      query.emplace_back(Queries[i][0] - 1, Queries[i][1] - 1, i);
    }

    sort(query.begin(), query.end(),
         [&](tuple<int, int, int> &a, tuple<int, int, int> &b) {
           int a_block = get<0>(a) / block_size;
           int b_block = get<0>(b) / block_size;
           if (a_block != b_block) {
             return a_block < b_block;
           }
           return get<1>(a) < get<1>(b);
         });

    int cl = 0, cr = -1, cnt = 0;
    unordered_map<int, int> freq;

    for (auto q : query) {
      int l = get<0>(q), r = get<1>(q);
      while (cl > l) {
        cl--;
        if (++freq[nums[cl]] == k) {
          cnt++;
        }
      }
      while (cr < r) {
        cr++;
        if (++freq[nums[cr]] == k) {
          cnt++;
        }
      }
      while (cl < l) {
        if (--freq[nums[cl]] == k - 1) {
          cnt--;
        }
        cl++;
      }
      while (cr > r) {
        if (--freq[nums[cr]] == k - 1) {
          cnt--;
        }
        cr--;
      }

      ans[get<2>(q)] = cnt;
    }

    return ans;
  }
};

//{ Driver Code Starts.
int main() {
  int tc;
  cin >> tc;
  while (tc--) {
    int n, q, k;
    cin >> n >> q >> k;
    vector<int> nums(n);
    for (int i = 0; i < n; i++)
      cin >> nums[i];
    vector<vector<int>> Queries;
    for (int i = 0; i < q; i++) {
      int l, r;
      cin >> l >> r;
      Queries.push_back({l, r});
    }
    Solution obj;
    vector<int> ans = obj.solveQueries(nums, Queries, k);
    for (auto i : ans)
      cout << i << "\n";
  }
  return 0;
}
// } Driver Code Ends

File: segment_tree/interesting_queries.py
# User function Template for python3
from collections import defaultdict
import math


class Solution:
    def solveQueries(self, nums, Queries, k):
        # Code here
        n = len(nums)
        nq = len(Queries)
        ans = [0] * nq
        block_size = int(math.sqrt(n)) + 1
        query = []

        for i in range(nq):
            query.append((Queries[i][0], Queries[i][1], i))

        query.sort(key=lambda x: (x[0] / block_size, x[1]))
        freq = defaultdict(int)

        cl, cr = 0, -1
        cnt = 0
        for q in query:
            l, r = q[0] - 1, q[1] - 1
            while cl > l:
                cl -= 1
                freq[nums[cl]] += 1
                if freq[nums[cl]] == k:
                    cnt += 1
            while cr < r:
                cr += 1
                freq[nums[cr]] += 1
                if freq[nums[cr]] == k:
                    cnt += 1
            while cl < l:
                freq[nums[cl]] -= 1
                if freq[nums[cl]] == k - 1:
                    cnt -= 1
                cl += 1
            while cr > r:
                freq[nums[cr]] -= 1
                if freq[nums[cr]] == k - 1:
                    cnt -= 1
                cr -= 1
            ans[q[2]] = cnt

        return ans


# {
# Driver Code Starts
# Initial Template for Python 3

if __name__ == "__main__":
    T = int(input())
    for i in range(T):
        n, q, k = input().split()
        n = int(n)
        q = int(q)
        k = int(k)
        nums = list(map(int, input().split()))
        Queries = []
        for _ in range(q):
            Queries.append(list(map(int, input().split())))
        obj = Solution()
        ans = obj.solveQueries(nums, Queries, k)
        for _ in ans:
            print(_)

# } Driver Code Ends
